### DISCUSSION TITLE
How to set event loop while testing app using pytest-asyncio

### DISCUSSION THREAD

**Discussion by PCanavelli** [fastapi/fastapi#8415] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import pytest
import sqlalchemy as sa

from asyncio import current_task
from fastapi import APIRouter
from fastapi import Depends
from fastapi import FastAPI
from fastapi.testclient import TestClient
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import async_scoped_session
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.declarative import DeclarativeMeta
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import sessionmaker

DB_URL = "SOME_DB_URL"
Base: DeclarativeMeta = declarative_base()
app = FastAPI()
test_router = APIRouter(prefix="test")
app.include_router(test_router)


class User(Base):

    __tablename__ = "users"
    user_id = sa.Column(sa.String, primary_key=True)
    name = sa.Column(sa.String)
    

def get_db_session() -> async_scoped_session:

    engine = create_async_engine(DB_URL)
    session_maker = sessionmaker(bind=engine, class_=AsyncSession)
    db_session = async_scoped_session(session_maker, scopefunc=current_task)

    return db_session


@test_router.post(path="/")
async def hello(
    user_id: str,
    db_session: async_scoped_session = Depends(get_db_session)
) -> str:
    
    query = select(User).where(User.user_id == user_id)
    result = await db_session.execute(query)
    user_name = result.scalars().first().name

    return f"hello {user_name}"


@pytest.fixture
def test_client():

    # some dependency overrides here, omitted for clarity as they don't impact the issue
    # app.dependency_overrides[get_db_session] = some_method_to_create_test_db_session
    
    yield TestClient(app)

    # reset overrides
    app.dependency_overrides = {}


async def test_post_success(test_client: TestClient):

    response = test_client.post(url="test/", params={"user_id": "some_id"})

    assert response == "hello some_user"
```

### Description

Hi everyone,

I need some help regarding the best way to test a FastAPI app interacting with a postgresql databse through asyncpg. More precisely, I am trying to swap psycopg2 for asyncpg, and rewriting all necessary methods and unit tests to make them async.

I'm using pytest-asyncio to unit test the newly-async code. Testing calls to the API is proving problematic: all async tests now fail with the dreaded `Future <Future pending cb=[Protocol._on_waiter_completed()]> attached to a different loop`.

From what I could find on pytest-asyncio's github, this kind of issue is not easily solvable on their end, and is usually addressed by creating an `event_loop` fixture and passing it to the app being tested with something like:

```python
@pytest.fixture(scope='session')
def event_loop():
    policy = asyncio.get_event_loop_policy()
    loop = policy.new_event_loop()
    yield loop
    loop.close()
```

However, this type of fix was suggested in cases that don't involve FastAPI/Uvicorn, and I am at a loss as to how to force-set FastAPI's event loop to use the test fixture.

Could anyone suggest the best way to do so, or an alternative method to solve this testing issue?

I've tried to submit a code snipped encapsulating the main details of my use case: the router structure, the db session generation and usage, and the test client fixture. Please feel free to request a rewrite of the example or additional info as needed.

Thanks in advance for your help

### Operating System

Linux

### Operating System Details

Debian

### FastAPI Version

fastapi==0.85.0

### Python Version

Python 3.9.2

**Comment by jgould22** [comment]
Did you try adding a `conftest.py` in the root of your test directory with the below content 

```python
import asyncio
import pytest

@pytest.fixture(scope="session")
def event_loop():

    """Create an instance of the default event loop for each test case."""
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()
```

I have had no issues with asyncpg / SQLA with the above

**Comment by PCanavelli** [comment]
Hi @jgould22, and thanks for your answer.

This solution is equivalent to the snippet described in my question, which unfortunately doesn't seem like a great option for two reasons:

1. I'm using the [pytest-postgresql](https://github.com/ClearcodeHQ/pytest-postgresql) fixture in most of my tests. This fixture is function-scoped by default, which conflicts with the use of a session-scoped event loop.
2. Relying on a session-scoped loop (and possibly a db as well to make everything mutually compatible) doesn't seem like good practice, as unit tests wouldn't be independent anymore  - so not actual unit tests.

**Reply by commentator8** [reply]
@PCanavelli I am running into the same issue, with the snipped above in use as well. Did you find a solution?

**Comment by eladgunders** [comment]
@PCanavelli @commentator8 
here is the link to my repo where I did it in `./todos/tests` directory: https://github.com/eladgunders/fastapi-todos
my reference was @Kludex repo: https://github.com/Kludex/fastapi-microservices

let me know if it was helpful and if you have any questions:)

**Reply by commentator8** [reply]
@eladgunders Thanks! I'll check it out.

**Comment by Harpsichord1207** [comment]
I have a SSE api, which returns an EventSourceResponse.
I set  event loop like that:
```
@pytest.fixture(scope='session')
def event_loop():
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()
```

The unit test is 
```
 @pytest.mark.asyncio
  async def test_count_notification(self):
      async with httpx.AsyncClient(app=app, base_url='http://test') as async_client:
          resp = await async_client.get('/count')
          assert resp.status_code == status.HTTP_200_OK
```

I still got an error:
```
E           RuntimeError: Task <Task pending name='sse_starlette.sse.EventSourceResponse.__call__.<locals>.wrap' coro=<EventSourceResponse.__call__.<locals>.wrap() running at E:\DevOpsPlatform\venv\lib\site-packages\sse_starlette\sse.py:240> cb=[TaskGroup._spawn.<locals>.task_done() at E:\DevOpsPlatform\venv\lib\site-packages\anyio\_backends\_asyncio.py:661]> got Future <Future pending> attached to a different loop
```

**Reply by Kludex** [reply]
https://pypi.org/project/httpx-sse/

**Reply by Harpsichord1207** [reply]
I tried use httpx-sse, still got `got Future <Future pending> attached to a different loop`。here's my code
```
        async with httpx.AsyncClient(app=app, base_url='http://test') as async_client:
            async with aconnect_sse(async_client, "GET", "/count") as event_source:
                async for sse in event_source.aiter_sse():
                    print(sse)
```

**Comment by elementary-charge** [comment]
Using AsyncClient from httpx library instead of fastapi TestClient works for me.
Example of client fixture

```
from httpx import AsyncClient

@pytest.fixture
async def client() -> AsyncClient:
    async with AsyncClient(app=app) as c:
        yield c
```

**Reply by sstraub-hmmc** [reply]
I was struggling with this, until I checked `asyncio.get_running_loop()` and noticed that `TestClient` always runs in a different event loop, no matter how I set up my tests. After switching to `AsyncClient` I was finally on the same event loop as the other async libraries like pymongo.

The `AsyncClient` constructor seems to have changed though; I had success with this one:

```python
from typing import AsyncGenerator, Any
import pytest
from httpx import AsyncClient, ASGITransport


@pytest.fixture(scope="session")
async def client() -> AsyncGenerator[AsyncClient, Any]:
    async with AsyncClient(transport=ASGITransport(app), base_url="http://localhost/") as ac:
        yield ac
```

Here's the relevant FastAPI documentation: https://fastapi.tiangolo.com/advanced/async-tests/

Running the fixture in session scope only works with the [right settings for pytest-asyncio](https://pytest-asyncio.readthedocs.io/en/stable/how-to-guides/change_default_test_loop.html) in `pyproject.toml`:

```toml
[tool.pytest.ini_options]
asyncio_mode = "auto"
asyncio_default_fixture_loop_scope = "session"
asyncio_default_test_loop_scope = "session"
```
